### PR TITLE
Improve this map

### PR DESCRIPTION
**Pull request by tristen**
This addition to infoControl listens for the existence of an element `mapbox-improve-map` and attaches the `_editLink` method to it.

### CODE REVIEW

**Comment by tmcw** [comment]
Plz test this with multiple maps, it will break.

**Comment by tmcw** [comment]
Holding on this until we have the server side pushed. And then I'm not really clear on why we're doing the event binding thingy when we could just rewrite the link url.

**Comment by lxbarth** [comment]
> And then I'm not really clear on why we're doing the event binding thingy when we could just rewrite the link url.

I don't care much either way, but @tristen can you sort this out not waiting for server side merge?

**Comment by tristen** [comment]
Sorry, I need more of an explanation here. I shouldn't be listening for the click event to re-write the url with the maps current coordinates?

**Comment by tmcw** [comment]
These tests need to cover when the map actually moves and the code rewrites the link - right now the whole `moveend` event binding is untested.

**Comment by tmcw** [comment]
(that should also cover moveend events with improvemap vs moveend events without the link)

**Comment by tmcw** [comment]
This binds an event `moveend` on the map object but does not unbind it when the control is removed.

**Comment by tmcw** [comment]
It also re-binds that event every time that the map is updated. Let's clean up the event binding situation here.

**Comment by tristen** [comment]
@tmcw I think I'm in a good place here and addressed your comments above.

File: src/info_control.js [hunk 1]
```diff
@@ -43,6 +43,7 @@ var InfoControl = L.Control.extend({
 
     onRemove: function(map) {
         map
+            .off('moveend', this._editLink, this)
```
Comment (tmcw) [inline src/info_control.js:46]: .off() doesn't have a third argument here

File: src/info_control.js [hunk 2]
```diff
@@ -91,26 +91,24 @@ var InfoControl = L.Control.extend({
...
-        '&lat=' + center.lat + '&lon=' + center.lng);
+    _editLink: function() {
+        var link = this._content.getElementsByClassName('mapbox-improve-map');
+        if (link.length && this._map._loaded) {
+            var center = this._map.getCenter();
+            var tilejson = this._tilejson || this._map._tilejson || {};
+            var id = tilejson.id || '';
+
+            for (var i = 0; i < link.length; i++) {
+                link[i].href = 'https://www.mapbox.com/map-feedback/#' + id + '/' + center.lng + '/' + center.lat + '/' + this._map.getZoom();
```
Comment (lxbarth) [inline src/info_control.js]: @tristen - any reason we echo the URL here? Let's just add to the URL instead of replacing it, no?

**Comment by lxbarth** [comment]
@tmcw - is this good to merge?

File: src/info_control.js [hunk 3]
```diff
@@ -91,26 +91,25 @@ var InfoControl = L.Control.extend({
...
-        '&lat=' + center.lat + '&lon=' + center.lng);
+    _editLink: function() {
+        var link = this._content.getElementsByClassName('mapbox-improve-map');
+        if (link.length && this._map._loaded) {
+            var center = this._map.getCenter();
+            var tilejson = this._tilejson || this._map._tilejson || {};
+            var id = tilejson.id || '';
+
+            for (var i = 0; i < link.length; i++) {
+                if (!this._editurl) this._editurl = link[i].href;
```
Comment (tmcw) [inline src/info_control.js]: Can you make this less brittle?

``` js
link[i].href = link[i].href.split('#') + '#' + id + '/' + center.lng + '/' + center.lat + '/' + this._map.getZoom();
```

**Comment by tmcw** [comment]
@lxbarth i haven't had the time to test it along with the tilestream-pro patch, and ideally that patch is merged first. The code looks okay except for ^ where it could be improved, but I'd need to see it actually working.

**Comment by lxbarth** [comment]
Awesome.

@tristen - please [address brittleness](https://github.com/mapbox/mapbox.js/pull/673#discussion_r10013638). This PR should be ready to be merged then.

@tmcw let's do voice on roll out order.

File: src/info_control.js (hunk 1 above)
Comment (mourner) [inline src/info_control.js:46]: Why not? I think there should be a third argument on all `off` calls if it was given in corresponding `on`.

**Comment by lxbarth** [comment]
@tmcw - after voice just now:
- Removed any irrelevant changes (mostly whitespace and debug stuff)
- Switched test to local server
- Simplified tests
- Tested whether infoControl works on old and new attribution (manually and in unit tests)
- Protected call to `getElementsByClassName` so IE8 or other older browsers don't choke
- Spot checked IE8 on browserstack

@tmcw - please review.

/cc @tristen

**Pull request merged**